applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.cpp

```diff
@@ -0,0 +1,165 @@
+            KRATOS_INFO("EstimateDtUtility") << "Setting minimum delta time " << mDtMin << " as current time step." << std::endl;
+            new_dt = mDtMin;
+        } else {
+            // Compute new Dt
+            new_dt = mCFL * current_dt / current_cfl;
+            // Limit max and min Dt
+            if (new_dt > mDtMax) {
+                new_dt = mDtMax;
+            } else if (new_dt < mDtMin) {
+                new_dt = mDtMin;
+            }
+        }
+
+        // Perform MPI sync if needed
+        new_dt = mrModelPart.GetCommunicator().GetDataCommunicator().MinAll(new_dt);
+
+        return new_dt;
+
+        KRATOS_CATCH("")
+    }
+
+    void EstimateDtUtility::CalculateLocalCFL(ModelPart& rModelPart)
+    {
+        KRATOS_TRY;
+
+        // Get the projected element size function according to the corresponding geometry
+        // Note that in here it is assumed that all the elements in the model part feature the same geometry
+        const auto& r_geom = rModelPart.ElementsBegin()->GetGeometry();
+        ElementSizeFunctionType projected_h_func = EstimateDtUtility::GetMinimumElementSizeFunction(r_geom);
+
+        // Calculate the CFL number in each element
+        const double current_dt = rModelPart.GetProcessInfo().GetValue(DELTA_TIME);
+        block_for_each(rModelPart.Elements(), [&](Element& rElement){
+            const double element_cfl = EstimateDtUtility::CalculateElementCFL(rElement, projected_h_func, current_dt);
+            rElement.SetValue(CFL_NUMBER, element_cfl);
+        });
+
+        KRATOS_CATCH("")
+    }
+
+    double EstimateDtUtility::CalculateElementCFL(
+        const Element &rElement,
+        const ElementSizeFunctionType& rElementSizeCalculator,
+        const double Dt)
+    {
+        // Calculate the midpoint velocity
+        const auto& r_geometry = rElement.GetGeometry();
+        const unsigned int n_nodes = r_geometry.PointsNumber();
+        array_1d<double,3> element_vel = r_geometry[0].FastGetSolutionStepValue(VELOCITY);
+        for (unsigned int i = 1; i < n_nodes; ++i) {
+            element_vel += r_geometry[i].FastGetSolutionStepValue(VELOCITY);
+        }
+        element_vel /= static_cast<double>(n_nodes);
+
+        // Calculate element CFL
+        const double h_min = rElementSizeCalculator(r_geometry);
+        const double elem_cfl = norm_2(element_vel) * Dt / h_min;
+
+        return elem_cfl;
+    }
+
+    typename EstimateDtUtility::ElementSizeFunctionType EstimateDtUtility::GetMinimumElementSizeFunction(const Geometry<Node<3>>& rGeometry)
+    {
+        ElementSizeFunctionType projected_h_func;
+        const auto geometry_type = rGeometry.GetGeometryType();
+        if (geometry_type == GeometryData::Kratos_Triangle2D3) {
+            projected_h_func = [&](const Geometry<Node<3>>& rGeometry){return ElementSizeCalculator<2,3>::MinimumElementSize(rGeometry);};
```

Comment:
```suggestion
            return [&](const Geometry<Node<3>>& rGeometry){return ElementSizeCalculator<2,3>::MinimumElementSize(rGeometry);};
```

cannot you do that?

Reply:
But isn't this equivalent to the `else if` statement?

Reply:
OK. Now I see your point. I think this yields a warning as there is one case without return.

Reply:
In any case I changed it to a `switch` statement.